### ISSUE TITLE
Pitfall: voxel array shape handedness must match `QuadCoordinateConfig` handedness

### ISSUE THREAD

**Issue by Inspirateur** [bonsairobo/block-mesh-rs#29]
## Issue
![image](https://github.com/bonsairobo/block-mesh-rs/assets/22746898/f639a4bc-f3e6-4d76-85bf-ec52093078cf)

Here we can clearly see that one face is missing in this mesh and I'm really struggling to understand why.

## Info
- The library has worked very well for me when I was using the visible_block_faces function, but since I switched to greedy meshing i sometimes get holes like this. 
- my chunk shape is Y first ! meaning columns are contiguous in the voxel buffer i'm passing to greedy_quads(), maybe that's the issue ?
- The holes are rare in natural terrain and behave *very* strangely  
(a hole can be toggled by changing a bloc that is not even connected to it)
- All my blocks have visibility "Opaque" except Air which is "Empty".
- The texture is hardcoded to basalt for debugging purposes so I know it's not a texturing issue

## Code snippet
```rust
    fn create_face_meshes(&self, chunk: ChunkPos, texture_map: &DashMap<(Block, FaceSpecifier), usize>) -> [Option<Mesh>; 6] {
        let padded_chunk_shape = YFirstShape::new_padded();
        
        // i don't think the problem can come from here given the screenshot
        let mut voxels = vec![Block::Air; padded_chunk_shape.size3];
        self.fill_padded_chunk(&mut voxels, chunk, &padded_chunk_shape);

        let mut buffer = GreedyQuadsBuffer::new(padded_chunk_shape.size3);
        greedy_quads(
            &voxels,
            &padded_chunk_shape,
            [0; 3],
            [CHUNK_S1+1; 3],
            &RIGHT_HANDED_Y_UP_CONFIG.faces,
            &mut buffer,
        );

        let num_quads = buffer.quads.num_quads();
        
        // we return 1 mesh per face type for efficient culling 
        [0, 1, 2, 3, 4, 5].map(|i| {
            let mut indices = Vec::with_capacity(num_quads * 6);
            let mut voxel_data: Vec<[u32; 2]> = Vec::with_capacity(num_quads * 4);
            let n = i as u32;
            let quad_face = &RIGHT_HANDED_Y_UP_CONFIG.faces[i];

            for quad in buffer.quads.groups[i].iter(){
                let mesh_positions = &quad_face.quad_mesh_positions(quad.into(), 1.0)
                    .map(|[x, y, z]| [x-1.0, y-1.0, z-1.0]);
                let uvs = quad_face.tex_coords(RIGHT_HANDED_Y_UP_CONFIG.u_flip_face, true, &quad);
                // usually chosen based on block but hardcoded here for debugging
                let texlayer = 1;
                let color = 0b111_111_111;
                let light = 0u32;
                indices.extend_from_slice(&quad_face.quad_mesh_indices(voxel_data.len() as u32));
                // packing voxel data into 2 u32
                voxel_data.extend(mesh_positions.into_iter().zip(uvs.into_iter()).map(|(pos, uv)| {
                    let [x, y, z] = [pos[0] as u32, pos[1] as u32, pos[2] as u32];
                    let [u, v] = uv;
                    let first = x | (y << 6) | (z << 12) | (n << 18) | (color << 23);
                    let u = u as u32;
                    let v = v as u32;
                    let second = u | (v << 6) | (light << 12) | (texlayer << 16);
                    [first, second]
                }));
            }
            if indices.is_empty() {
                return None;
            }

            Some(
                Mesh::new(
                    PrimitiveTopology::TriangleList,
                    RenderAssetUsages::default(),
                )
                .with_inserted_attribute(ATTRIBUTE_VOXEL_DATA, voxel_data)
                .with_inserted_indices(Indices::U32(indices)),
            )
        })
    }
}
```

So yeah, I assume it's not a bug from the library but I don't know where else to look at this point...

**Comment by Inspirateur** [comment]
## Further debug information
In an attempt to find a minimal example, I've reduced my chunk sizes to 4 and reproduced the bug in an otherwise empty chunk, like so:
![image](https://github.com/bonsairobo/block-mesh-rs/assets/22746898/25cbc33f-6c7f-48fe-9d65-7ddb5d7194f1)

I printed out the content of `voxels` which is the voxel buffer fed into greedy_quads(), and formatted it for easier viewing:
```
[
empty slice 

empty slice 

empty slice 

0, 0, 0, 0, 0, 0, 
0, 0, 0, 0, 0, 0, 
0, 0, 1, 1, 0, 0, 
0, 0, 1, 0, 0, 0, 
0, 0, 0, 0, 0, 0, 
0, 0, 0, 0, 0, 0, 

empty slice 

empty slice 
]
```
(0 = Air, 1 = Stone)
My shape is Y first (the column are contiguous) so each line corresponds to a column (in bottom -> up),
each group of 6 columns correspond to a vertical slice, and 6 slices make a padded 4x4x4 chunk (which becomes 6x6x6 with padding)

`empty slice` is a block of 6x6 '0'  
In the middle we can see our 3 blocs exactly as I placed them, so the buffer I'm providing is correct, as expected

**Comment by Inspirateur** [comment]
Since I'm starting to think that the issue is a combination of greedy meshing + my Y-first chunk shape I'll paste my chunk shape here for further info:
```rust
use block_mesh::ndshape::Shape;

use super::CHUNK_S1;

pub struct YFirstShape {
    pub size1: usize,
    pub size2: usize,
    pub size3: usize,
    pub lod: usize
}

impl YFirstShape {
    pub fn new() -> Self {
        let size1: usize = CHUNK_S1;
        Self {
            size1,
            size2: size1*size1,
            size3: size1*size1*size1,
            lod: 1
        }
    }

    pub fn new_padded(lod: usize) -> Self {
        let size1: usize = CHUNK_S1/lod+2;
        Self {
            size1,
            size2: size1*size1,
            size3: size1*size1*size1,
            lod
        }
    }

    #[inline]
    pub fn linearize(&self, x: usize, y: usize, z: usize) -> usize {
        y + x * self.size1 + z * self.size2
    }
}

impl Shape<3> for YFirstShape {
    type Coord = u32;

    fn size(&self) -> Self::Coord {
        self.size3 as u32
    }

    fn usize(&self) -> usize {
        self.size3
    }

    fn as_array(&self) -> [Self::Coord; 3] {
        [self.size1 as u32; 3]
    }

    fn linearize(&self, p: [Self::Coord; 3]) -> Self::Coord {
        self.linearize(p[0] as usize, p[1] as usize, p[2] as usize) as u32
    }

    fn delinearize(&self, mut i: Self::Coord) -> [Self::Coord; 3] {
        let z = i / self.size2 as u32;
        i -= z * self.size2 as u32;
        let x = i / self.size1 as u32;
        let y = i % self.size1 as u32;
        [x, y, z]
    }
}
```
Maybe I misunderstood something when building my shape ? either way I think this issue will serve as good documentation for ppl in the same case as me.

**Comment by bonsairobo** [comment]
I'll try to find some time to look at this more later this week.

BTW you don't need to implement your own `Shape<3>`. You only need to permute the coordinates you pass to an existing shape implementation.

**Comment by Inspirateur** [comment]
> BTW you don't need to implement your own `Shape<3>`. You only need to permute the coordinates you pass to an existing shape implementation.

Thanks for the info ! I'll try and do that tomorrow :)

**Comment by bonsairobo** [comment]
> BTW you don't need to implement your own Shape<3>. You only need to permute the coordinates you pass to an existing shape implementation.

Oh actually I guess the issue is that it's the block mesh algos using the shape which would need to do the permutation. You could at least make your custom shape simpler by just wrapping an existing shape and doing the permutation in `linearize` and `delinearize`.

**Comment by Inspirateur** [comment]
> Oh actually I guess the issue is that it's the block mesh algos using the shape which would need to do the permutation. You could at least make your custom shape simpler by just wrapping an existing shape and doing the permutation in `linearize` and `delinearize`.

So i tried to write my shape like this
```rust
use block_mesh::ndshape::{RuntimeShape, Shape};
use super::CHUNK_S1;

pub struct YFirstShape {
    shape: RuntimeShape::<u32, 3>,
    pub lod: usize
}

impl YFirstShape {
    pub fn new() -> Self {
        Self {
            shape: RuntimeShape::<u32, 3>::new([CHUNK_S1 as u32; 3]),
            lod: 1
        }
    }

    pub fn new_padded(lod: usize) -> Self {
        let size1: usize = CHUNK_S1/lod+2;
        Self {
            shape: RuntimeShape::<u32, 3>::new([size1 as u32; 3]),
            lod
        }
    }

    pub fn linearize(&self, x: usize, y: usize, z: usize) -> usize {
        self.shape.linearize([y as u32, x as u32, z as u32]) as usize
    }
}

impl Shape<3> for YFirstShape {
    type Coord = u32;

    fn size(&self) -> Self::Coord {
        self.shape.size()
    }

    fn usize(&self) -> usize {
        self.shape.usize()
    }

    fn as_array(&self) -> [Self::Coord; 3] {
        self.shape.as_array()
    }

    fn linearize(&self, [x, y, z]: [Self::Coord; 3]) -> Self::Coord {
        self.shape.linearize([y, x, z])
    }

    fn delinearize(&self, i: Self::Coord) -> [Self::Coord; 3] {
        let [x, y, z] = self.shape.delinearize(i);
        [y, x, z]
    }
}
```

Did I do it wrong ? it wasn't much simpler 
(and it didn't fix the issue 😓)

**Comment by bonsairobo** [comment]
@Inspirateur If you could create a minimal repro example that would help me jump right into debugging.

**Comment by Inspirateur** [comment]
> @Inspirateur If you could create a minimal repro example that would help me jump right into debugging.

Yep, planning on doing that today; will update you when it's done :)

**Comment by Inspirateur** [comment]
@bonsairobo so I reproduced the bug starting from your wireframe example:
![image](https://github.com/bonsairobo/block-mesh-rs/assets/22746898/cf7bb393-e3c3-476b-9c6a-2e48a239357a)

I basically copied your code, keeping only the greedy mesh, changed the sphere voxel buffer to this:
```rust
    let mut samples = [BoolVoxel(false); YFirstShape::SIZE3];
    samples[122] = BoolVoxel(true);
    samples[123] = BoolVoxel(true);
    samples[128] = BoolVoxel(true);
```

and using this chunk shape:
```rust
use block_mesh::ndshape::Shape;

pub struct YFirstShape;

impl YFirstShape {
    pub const SIZE1: usize = 6;
    pub const SIZE2: usize = YFirstShape::SIZE1*YFirstShape::SIZE1;
    pub const SIZE3: usize = YFirstShape::SIZE2*YFirstShape::SIZE1;     
}

impl Shape<3> for YFirstShape {
    type Coord = u32;

    fn size(&self) -> Self::Coord {
        YFirstShape::SIZE3 as u32
    }

    fn usize(&self) -> usize {
        YFirstShape::SIZE3
    }

    fn as_array(&self) -> [Self::Coord; 3] {
        [YFirstShape::SIZE1 as u32; 3]
    }

    fn linearize(&self, p: [Self::Coord; 3]) -> Self::Coord {
        // y and x are swapped here
        p[1] + p[0] * YFirstShape::SIZE1 as u32 + p[2] * YFirstShape::SIZE2 as u32
    }

    fn delinearize(&self, mut i: Self::Coord) -> [Self::Coord; 3] {
        let z = i / YFirstShape::SIZE2 as u32;
        i -= z * YFirstShape::SIZE2 as u32;
        let y = i / YFirstShape::SIZE1 as u32;
        let x = i % YFirstShape::SIZE1 as u32;
        // y and x are swapped here
        [y, x, z]
    }
}
```

If i undo the swapping between Y and X in the Chunk shape I get this correct result instead:
![image](https://github.com/bonsairobo/block-mesh-rs/assets/22746898/a1eee4f0-6fda-4a66-ba96-7bb20286960e)

Here's a zip of the code if you need it:
[mesh-bug.zip](https://github.com/bonsairobo/block-mesh-rs/files/14932905/mesh-bug.zip)

**Comment by bonsairobo** [comment]
@Inspirateur I have a solution for you. This was based on a hunch that turned out to be correct, and I haven't really validated it by reading the code (which is well evicted from my brain cache).

TL;DR: If you switch your `YFirstShape` to index by `[y, z, x]` instead of `[y, x, z]`, then the issue goes away.

# Explanation

By swapping two coordinates instead of [rotating](https://en.wikipedia.org/wiki/Cyclic_permutation) all of them, you are implicitly switching from a right-handed coordinate frame to a left-handed one. The `RIGHT_HANDED_Y_UP_CONFIG` is not compatible with a left-handed coordinate frame. ([relevant math jargon](https://en.wikipedia.org/wiki/Parity_of_a_permutation))

If you don't want to change your array layout, then you would need to define a new `QuadCoordinateConfig`.

**Comment by bonsairobo** [comment]
You might be best served by defining a new `QuadCoordinateConfig` because that will give you the correct quad coordinates immediately instead of needing to transform the quad coordinates to invert your axis permutation.

**Comment by bonsairobo** [comment]
As far as resolving this issue, I think there is no bug, but we could definitely improve documentation around this pitfall.

**Comment by Inspirateur** [comment]
aaah this totally makes sense ! And making the switch is easy for me so I'll do just that ! I think this issue is good enough of a documentation for now, thanks a lot :)

i'll try it tomorrow and tell you how it goes.

**Comment by Inspirateur** [comment]
Mh, simply changing my shape to Y, Z, X didn't work, maybe I need to define a custom QuadCoordinate too ? but I'm not sure how to define the correct one for my shape ...

**Comment by Inspirateur** [comment]
I also reproduced it with the minimal example: 
![image](https://github.com/bonsairobo/block-mesh-rs/assets/22746898/348faf7c-63c5-4923-ac1e-c74b490aa318)
Here I changed the shape to Y, Z, X and set the blocks 122, 128, 164 instead (and left the RIGHT_HANDED_Y_UP_CONFIG as is)

**Comment by Inspirateur** [comment]
I also tried to fiddle with the AxisPermutation in the QuadCoordinateConfig (with and without my custom shape) but couldn't find a working combination :(

**Comment by Inspirateur** [comment]
So a solution that worked is to not use a custom shape or a custom quad config, but simply map the positions back to undo the axis swap, like so:
```rust
positions.extend_from_slice(&face.quad_mesh_positions(&quad, 1.0).map(|[y, z, x]| [x, y, z]));
```

I also had to use this custom QuadCoordinateConfig:
```rust
const Y_FIRST_RIGHT_HANDED_Y_UP_CONFIG: QuadCoordinateConfig = QuadCoordinateConfig {
    faces: [
        OrientedBlockFace::new(-1, AxisPermutation::Zxy),
        OrientedBlockFace::new(-1, AxisPermutation::Xzy),
        OrientedBlockFace::new(-1, AxisPermutation::Yzx),
        OrientedBlockFace::new(1, AxisPermutation::Zxy),
        OrientedBlockFace::new(1, AxisPermutation::Xzy),
        OrientedBlockFace::new(1, AxisPermutation::Yzx),
    ],
    u_flip_face: Axis::X,
};```

and had to swap the uvs like this for the left/right faces to have correct orientation
```rust
if face == Face::Left {
    uvs[0].swap(0, 1);
    uvs[1].swap(0, 1);
    uvs[2].swap(0, 1);
} else if face == Face::Right {
    uvs.swap(0, 1);
    uvs.swap(2, 3);
    uvs[0].swap(0, 1);
    uvs[1].swap(0, 1);
    uvs[2].swap(0, 1);
}
```

which is pretty ugly but at least it works 🫠
I think imma go with that and consider this issue closed.
